packages/gestalt/src/Dropdown.js

```diff
@@ -0,0 +1,239 @@
+// @flow strict
+import React, {
+  Children,
+  cloneElement,
+  useState,
+  useRef,
+  type Node,
+} from 'react';
+import PropTypes from 'prop-types';
+import Box from './Box.js';
+import Flyout from './Flyout.js';
+import Layer from './Layer.js';
+import DropdownItem from './DropdownItem.js';
+import DropdownSection from './DropdownSection.js';
+import DropdownContext from './DropdownContextProvider.js';
+import { type Indexable } from './zIndex.js';
+import handleContainerScrolling, {
+  type DirectionOptionType,
+} from './utils/keyboardNavigation.js';
+import { type AbstractEventHandler } from './AbstractEventHandler.js';
+import { ESCAPE, SPACE, TAB, ENTER, UP_ARROW, DOWN_ARROW } from './keyCodes.js';
+import { type OptionObject } from './MenuOption.js';
+
+type Props = {|
+  anchor?: ?HTMLElement,
+  children: Node,
+  headerContent?: Node,
+  idealDirection?: 'up' | 'right' | 'down' | 'left',
+  onDismiss: () => void,
+  onSelect?: AbstractEventHandler<
+    SyntheticKeyboardEvent<HTMLElement> | SyntheticMouseEvent<HTMLElement>,
+    {| item: ?OptionObject |}
+  >,
+  zIndex?: Indexable,
+|};
+
+const KEYS = {
+  UP: -1,
+  DOWN: 1,
+  ENTER: 0,
+};
+export default function Dropdown({
+  anchor,
+  children,
+  headerContent,
+  idealDirection = 'down',
+  onDismiss,
+  onSelect,
+  zIndex,
+}: Props): Node {
+  const flattenedChildren = Children.toArray(children).reduce(
+    (accumulatedChildren, currentChild) => {
+      const {
+        props: { children: currentItemChildren },
+        type: { displayName },
+      } = currentChild;
+      if (currentItemChildren && displayName === 'DropdownSection') {
+        return [...accumulatedChildren, ...currentItemChildren];
+      }
+      if (displayName === 'DropdownItem') {
+        return [...accumulatedChildren, currentChild];
+      }
+      // eslint-disable-next-line no-console
+      console.warn(
```

Comment:
I wonder if we could throw an Error instead. warning vs error?

Reply:
Once we get a chance to chat with the others about typing this cmp (the children prop is not yet properly typed b/c of errors that come up), this will go away anyway and we'll rely on that